File: Leetcode/47. Permutations II.py
class Solution:
    def permute(self,res :List[List[int]],start,nums):
        if start==len(nums):
            res.append(nums[:])
            return
        s=set()
        for i in range(start,len(nums)):
            if nums[i] not in s:
                s.add(nums[i])
                nums[i],nums[start]=nums[start],nums[i]
                self.permute(res,start+1,nums)
                nums[i],nums[start]=nums[start],nums[i]
        return 
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res=[]
        self.permute(res,0,nums)
        return res
        

File: Leetcode/47. Permutations II.cpp
class Solution {
public:
    void permute(vector<int>& nums,int start,vector<vector<int>>& res){
        if(start==nums.size()){
            res.push_back(nums);
            return;
        }
        unordered_set<int> s;
        for(int i=start;i<nums.size();i++){
            if(s.count(nums[i])==0){
                s.insert(nums[i]);
                swap(nums[start],nums[i]);
                permute(nums,start+1,res);
                swap(nums[start],nums[i]);
            }
        }
        return;
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        permute(nums,0,res);
        return res;
    }
};
